### ISSUE TITLE
Possibly another bug?

### ISSUE THREAD

**Issue by chiru1205** [confluentinc/confluent-kafka-dotnet#226]
Noticed on a Visual Studio rebuild of my application every time. App is hosted on IIS.

I get a JustInTime Debugger message and an event is registered in the Event Viewer
![image](https://user-images.githubusercontent.com/16243876/27146653-a5c99ada-50ff-11e7-9528-fd1be0e3a8cf.png)

I did collect data to analyze whats going on . Here is what the analysis tool gave me.
![image](https://user-images.githubusercontent.com/16243876/27146693-ce748a1c-50ff-11e7-88bf-9bb66ee6b407.png)

Just reporting it her, I'm not quite sure if this is a bug or something with my local machine.

**Comment by mhowlett** [comment]
don't have any good ideas on what's going on here - would need to see the code that might have caused it. At this point not going to take any action on this, but good to have in the back of my mind - thanks!

**Comment by chiru1205** [comment]
No worries. Just reporting it here as I test the client with a load of data.

**Comment by chiru1205** [comment]
@mhowlett @treziac . I thought this is not an issue ,but noticed overnight all my QA app pools shutdown and services became unavailable. Same errors in the event log. Seems like there is something going on.  I cannot figure what it is. As I dig through the crash dump file. Everything is pointing back to librdkafka.dll in memory.

![image](https://user-images.githubusercontent.com/16243876/27198076-5cf91a9c-51d6-11e7-9215-76e5f2f053a0.png)

A more detailed analysis of the crash dump file gave me this.
![image](https://user-images.githubusercontent.com/16243876/27198116-8925ba62-51d6-11e7-9f87-651360e10601.png)

Seems like librdkafka.dll is module is completely unloaded from memory. Wonder what would be the reason. This is concerning. I'm not sure if anyone encountered this before.

**Comment by chiru1205** [comment]
This crash appears to be happening when apppools are recycled(scheduled at specific time). I could recreate it locally when I rebuild the application. Apppool crashes.

![image](https://user-images.githubusercontent.com/16243876/27301023-6c35afc4-54f7-11e7-9e71-9d25a16036fb.png)


Could this be related to Locking of dll's ? [https://github.com/confluentinc/confluent-kafka-dotnet/issues/91](url)

**Comment by treziac** [comment]
Right, forgot about it
See #91 For a workaround. Maybe not anymore relevant with the changes we did - I also think calling freelibrary when recycling should do it, but it means you have to flush and wait for it to stop, so some config may be needed on your iis. I can confirm it worked without issue on our side to copy the dll in system folder.

**Comment by chiru1205** [comment]
@treziac , I did the workaround as suggested in https://github.com/confluentinc/confluent-kafka-dotnet/issues/91 , but still same issue. I might try copying to sys folder and see if that works for now. I don't know where to add flush in a class library because I never dispose the object(singleton). Is there a way to add flush in a class librabry? 

Can you elaborate on config setting you mentioned for iis. Your help is much appreciated

**Comment by chiru1205** [comment]
For me this is a serious bug as we cannot move our code to staging because of this issue. I tried all workarounds as mentioned in #91 , even copying the dll's to system32 folder. Still noticed the same issue where there is CLR exception(second chance) and apppool shutsdown. I'm surprised no one reported this.


Here is the exception I got after adding the files to system32 and removing build references to librdkafka. Still same error message

![image](https://user-images.githubusercontent.com/16243876/27308942-effcfb4a-5516-11e7-96d2-82d5c6217382.png)

**Comment by mhowlett** [comment]
so it looks like it's actually calling into the librdkafka code from the stack trace on the left? (if so that's good, should be able to work out a solution sure). Are the other errors you're highlighting simply because it can't load debug symbols? perhaps try putting lots of logging in your code to try and determine exactly where it's crashing and the parameters to the function call.

**Comment by chiru1205** [comment]
I'll debug more and will eventually get to the stack trace. Exceptions on the right are more related to modules unload which are handled in the CLR. Second hand exceptions are a really concern which are responsible for app crashes atleast in my case. Seems me like in the  middle of some metadata request calls  , dll is unloaded and blows up. May be @edenhill can help us here.

**Comment by treziac** [comment]
You may need to Flush and Dispose the singleton in your Application_End (in Global.asax) and call FreeLibrary, this is what I did on my side (with RdKafka which has basically the same code for this, I'm not using IIS anymore)

What I think is happening is that your worker is closed after recycling but did not finish with librdkafka process (still in flight metadata request), and when callback is called, the deliveryHandler has been release when w3p was stopped, thus the exception.
When pool is recycled, it won't receive any more requests, so this should give you time to treat all inflight request. Calling dispose will make sure there won't be aymore metdata request done for example 

You should make the shutdownTimeLimit as long as your message timeout, to be sure you wil have time to dispose (Dispose is blocking and wait for all in-flight request to complete). This is set in the advanced property of your app pool.

If you can't wait that long, you will have to cancel all inflight request. See here:
https://github.com/edenhill/librdkafka/wiki/Proper-termination-sequence#speeding-up-termination
Never tried it my side.

I may be wrong, I'm not totally comfortable with IIS. But it should be easy to add some logs when Application_End is called and see if this is related

**Comment by chiru1205** [comment]
@treziac @mhowlett . I have fixed the issue by disposing producer and calling free library on Application_End. Thank you so much for your input

**Comment by treziac** [comment]
Great! Will add it in the wiki or in a webapi sample when I have time. Thanks again for feedback ;)

Note there is still the issue if timeout occurs while disposing and you don't give 'ough time for recycling (broker not available - we need to provide a purge mechanism). If you want I will inform you here when it's available, but it has to be implemented in librdkafka first and I don't know the priority on this. But it's a narrow case ;) (but necesseray for at least once scenario I think)
Also will try to make the loadlibrary more easy. Did you finally used the doll in x86/x64, in bin or in a custom location? (I had to use a different location because our deployment process needed to delete whole folder without letting time for disposing, this might not be your case)

**Comment by chiru1205** [comment]
Or I can provide a sample if you need one.   What I have done with x64/x86 dll's is
1. Remove the build entry for librdkafka redist from my framework project(csproj) file. This will help is not creating the x64/x86 folders in bin folder. Note that if the folders are created previously then they will still  be in bin but  just empty.
2. I have created a folder in my service project because its easy for me to manage them there.This folder and files are added to the solution , that way deployment is seamless without any manual copy.
3. In Application_start I load the lib's, used your code sample from #91 
4. In Applictation_end I actually call framework static to dispose and flush an internal producer and after that I unload lib's 

I have also tried using a destructor on my producer wrapper but it doesn't work . I got threadabort exceptions which is almost critical like the previous one I have encountered and app pool shutdown still happened.

Using Application_end I think is a clean way to dispose all internal handles. In addition to all this I have updated app pool shutdown time to 3 minutes just to give more time

**Comment by chiru1205** [comment]
And Yes let me know when purge is added to API . All this additional coding can be avoided. Also let me know if there is a better way to do loadlibrary and free library. Thank you so much for all your help.

**Comment by chiru1205** [comment]
@treziac , @mhowlett  . Well my happiness didn't last long. Our AppPools are shutting down automatically again even with all these fixes you guys suggested. Collected full dump and on further analysis. This is exactly why our apppools are shutting down. Here is the link
https://blogs.msdn.microsoft.com/dsvc/2013/10/02/the-curious-case-of-bootup_exception_complus-0xc0020001-with-mixed-mode/

**Comment by chiru1205** [comment]
The issue is very easily recreated  with application deployed to IIS and the overlap recycle is enabled on the apppool. I'm to the point where I'll have to move on from this .NET client and rely on kafka Rest Proxy.

**Comment by treziac** [comment]
I didn't have problem on my own from memory (note that my native dll where not deploy with the application but in an other directory where I loaded them

Two think you can try:
1) You say you call LoadLibrary from Application_Start. We also call LoadLibraryEx in our code (https://github.com/confluentinc/confluent-kafka-dotnet/blob/master/src/Confluent.Kafka/Impl/LibRdKafka.cs#L69) perhaps you should try to call FreeLibrary twice (once for your code, once for us)
2) Try to put the library in an other folder than your application pool. I forced C:/kafka to contain librdkafka/zlib on our environments and loaded from there, and desactivated the copy from the dll in our application. This worked well as far as I know
3) Be sure, when calling flush, that it return zero (meaning there is no more message waiting in the queue) If it does not return zero, this means you still have in flight message which could try to call librdkafka code by callbacks
4) Use this https://github.com/edenhill/librdkafka/wiki/Proper-termination-sequence#speeding-up-termination Never tried it on my side

I will try it on my side to reproduce and apply those on my side, but I don't think I will be able to give proper answer before this we, and owner is currently on vacation.

In all cases, thanks for feedback!

**Comment by chiru1205** [comment]
Thank you I'll try your recommendations. Will Keep you posted.

**Comment by chiru1205** [comment]
![image](https://user-images.githubusercontent.com/16243876/27620883-5a49c524-5b92-11e7-9e3d-5974baabaa19.png)

Here is a more detailed error

**Comment by treziac** [comment]
Are you calling listgroup on your side?

I don't manage to replicate on iis10/win10... What are your iis and os version? Can you provide a minimal example easily reproducing? (Notably how you publish your asp.net project and iis configs?)

**Comment by chiru1205** [comment]
@treziac . Here is the finalizer thread call stack. In the image I have highlighted two areas. It appears after rd_kafka_destroy , rd_kafka_metadata is called which is calling back into our application via delegate. Our worker process shutdown is 90 seconds if this time is elapsed process is shutdown and appdomain is unloaded. At this point still that unmanaged thread is being run( as it cant be cleaned by CLR)  and trying to talk to the process which is already shutdown by IIS. 

![image](https://user-images.githubusercontent.com/16243876/27650027-3f5efad8-5bf9-11e7-8344-256d2a1c9377.png)



> Note that issue occurs only during scheduled recycle.

Issue is gone when I increased the ShutDown timelimit to 5 minutes instead of 90 seconds. Locally I was able to recreate it by changing the shutdown limit to 10 seconds or 20 seconds on the AppPool.

On the webserver we have a webgarden scenario, our apppool supports 10 worker processes with shutdown time of 90 seconds. So there is a chance that during recycle we have one or more worker process running to process request. 

I can't keep increasing the timelimit unless I know what is really going on.  Not sure about why metadata call is made after destroy of the handle.

I have also timed dispose call of the producer instance , takes roughly 0.5 to 1 second. 

My message time out in the configuration is set to 30 seconds. 

I'll put together a sample and upload it here.

**Comment by chiru1205** [comment]
@mhowlett @treziac . Sorry to reply this late on the issue. I took time off from work for a bit.
 I finally figured out what the issue is and resolved it on my end. Here is the list of steps to recreate the issue I have encountered.
1. Create a process where either Producer or consumer instance(s) are not disposed (i.e, flush is never called and dispose is not called either).
2. Host this process in IIS.
3. In our stage environment, each app pools allows maximum of 10 worker processes to handle requests(i.e, in a high load situation we have 10 process instances serving requests) and our Rapid File Protection in IIS has maximum failures set to 5. If the producer or consumer is not disposed , during recycle of app pool, each worker process will get a domain unloaded exception( this is a second chance exception and not handled in the Code). Because RFP maximum failure is set to 5 , once 5 of the 10 worker processes encounter this unhandled exception, app pool is shutdown automatically. Here is a screenshot from my local, my app pool has 3 worker process and RFP max failure is set to 1.
![image](https://user-images.githubusercontent.com/16243876/28271284-bd03fdac-6acd-11e7-8126-13e5d39c2dd2.png)

I hope this information will help somebody.
